internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala

```diff
@@ -937,7 +937,7 @@ private final class AnalysisCallback(
         if (!writtenEarlyArtifacts) // writing implies the updates merge has happened
           mergeUpdates() // must merge updates each cycle or else scalac will clobber it
       }
-      incHandler.completeCycle(invalidationResults, getAnalysis)
+      incHandler.completeCycle(None, getAnalysis)
```

Comment:
I see that this change may fix the particular bug regarding Scala / Java mixed situation, but this particular change doesn't feel right. Isn't this called once as the name suggest?

Reply:
Would you mind to elaborate on how `getCycleResultOnce` being called once is significant?

I would like to learn more about why the change may not be the right thing to do and investigate alternative approaches.

Reply:
Generally, the code in this area was contributed in https://github.com/sbt/zinc/pull/744 to implement build pipelining. I was involved in the code rewrite process to bring this into Zinc, but it's been a long while and I barely understood it back then.

The reason why the change didn't sit right with me (sorry about this vagueness) is because `completeCycle` is defined as:

```scala
    /**
     * Merge latest analysis as of analyzer into pruned previous analysis and inform file manager.
     */
    def completeCycle(
        prev: Option[CompileCycleResult],
        partialAnalysis: Analysis
    ): CompileCycleResult
```

So when there was a previous result either via API phase, or from the dependency phase, simply dropping this information would have some breakage down the road. If we're going to pretend there was no previous `CompileCycleResult`, my guess is we should do so elsewhere, like inside of `completeCycle(...)`.

>  Would you mind to elaborate on how `getCycleResultOnce` or `completeCycle` being called once is significant?

Maybe I confused myself the once being important to the timing, but I did find an interesting comment made by 2020 me - https://github.com/sbt/zinc/pull/744#issuecomment-623628523:

> If we try to pipeline mixed Java/Scala subprojects, the Analysis information created by JavaAnalyze will not be available to downstream subprojects. Instead it will use the Analysis information created via `-Ypickle-java + xsbt-dependency` phase etc. I'm not sure if we lose some accuracy because of this.

and in https://github.com/sbt/sbt/pull/5703 I implemented `compileJava` task in sbt, which is called only if pipelining is enabled. You wrote:

> [invalidationResults] is not updated again after Java analysis, hence `invalidationResults` contains incorrect information for Java sources.

so I guess we do have loss of information when there's a mix of Scala and Java, and it's not pipelined. Since we're the one running [AnalyzingJavaCompiler](https://github.com/sbt/zinc/blob/develop/zinc/src/main/scala/sbt/internal/inc/javac/AnalyzingJavaCompiler.scala) I wonder if it's possible to somehow pass in the right information to AnalysisCallback?

Reply:
Thanks for your explanation, Eugene!

> So when there was a previous result either via API phase, or from the dependency phase, simply dropping this information would have some breakage down the road. If we're going to pretend there was no previous CompileCycleResult

If we discard `invalidationResult` when calling `completeCycle`, then `completeCycle` would just call `mergeAndInvalidate`, which reconstructs the most up-to-date analysis from `partialAnalysis` (being `getAnalysis`) and pruned analysis, and compute `CompileCycleResult` using the reconstructed analysis.

Since the `partialAnalysis` contains information extracted from `API` and `Dependency` phase, I thought  `completeCycle(None, getAnalysis)` should still correctly account for information extracted from `API` and `Dependency` phase.

> Since we're the one running [AnalyzingJavaCompiler](https://github.com/sbt/zinc/blob/develop/zinc/src/main/scala/sbt/internal/inc/javac/AnalyzingJavaCompiler.scala) I wonder if it's possible to somehow pass in the right information to AnalysisCallback?

Correct me if I misinterpreted this comment, but if the intention of the comment is to making sure that `invalidationResults` correctly accounts Java analysis in the first place, I think we still have to call `mergeAndInvalidate`, which basically yields a very similar solution (instead of calling `mergeAndInvalidate` in `completeCycle`, we call it before in `AnalyzingJavaCompiler`.) But I can see the benefit of that solution (it maintains the invariant that `invalidationResults` is always fresh).

---

This might be out of scope of this PR and might be a real stretch. But... can we get rid of `invalidationResults` altogether? When we need to know about invalidation, we can always just reuse partial and pruned analysis and extract the result from calling `mergeAndInvalidate`. In fact, other than the one call to `incHandler.completeCycle(invalidationResults, getAnalysis)`, the exact content of `invalidationResults` is fetched nowhere. Rather various usage of `invalidationResults`  merely checks if `invalidationResults` is empty or not. Less variable equals to less bug.

Reply:
My guess is that `invalidationResults` var is/was necessary to hand of these information mid-way through the last cycle.

![pipelining](https://github.com/sbt/zinc/assets/184683/ad7edead-a1c9-4a89-957b-1f234cfb1c0a)

I guess one light change you could try is adding `earlyOutput.isDefined &&` to
https://github.com/sbt/zinc/blob/1c809a60a11a710aa4b016aa28f7e8a44da68a98/internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala#L1093
that way we can avoid calling `incHandler.mergeAndInvalidate(a, false)` needlessly if we're not gonna use it, and `invalidationResults` won't be assigned.